### ISSUE TITLE
`tests/test_least_squares.py::test_least_squares[rosenbrock-minimum2-init2-args2-solver15]` fails with jax==0.7.1

### ISSUE THREAD

**Issue by GaetanLepage** [patrick-kidger/optimistix#160]
Context: [bumping jax from 0.6.2 to 0.7.1 in nixpkgs](https://github.com/NixOS/nixpkgs/pull/427588).

Since updating jax to 0.7.1, I noticed that the following test was failing:
```
_________ test_least_squares[rosenbrock-minimum2-init2-args2-solver15] _________
[gw15] linux -- Python 3.12.11 /nix/store/18l19gnl8dq75v6knghbcgkrikmpc0yv-python3-3.12.11/bin/python3.12

solver = OptaxMinimiser(
  optim=GradientTransformationExtraArgs(
    init=_Closure(
      fn=<equinox.internal._closure_to_pyt...        )
        ),
      )
    )
  ),
  rtol=1e-08,
  atol=1e-08,
  norm=<function max_norm>,
  verbose=frozenset()
)
_fn = <function rosenbrock at 0x7fff5c3113a0>
minimum = Array(0., dtype=float64, weak_type=True)
init = (Array([[1.5, 1.5, 1.5, 1.5],
       [1.5, 1.5, 1.5, 1.5]], dtype=float64), {'a': Array([[[1.5, 1.5],
        [1.5, 1.5],
        [1.5, 1.5]],

       [[1.5, 1.5],
        [1.5, 1.5],
        [1.5, 1.5]]], dtype=float64)}, ())
args = Array(1., dtype=float64, weak_type=True)

    @pytest.mark.parametrize("solver", least_squares_optimisers)
    @pytest.mark.parametrize("_fn, minimum, init, args", least_squares_fn_minima_init_args)
    def test_least_squares(solver, _fn, minimum, init, args):
        atol = rtol = 1e-4
        has_aux = random.choice([True, False])
        if has_aux:
            fn = lambda x, args: (_fn(x, args), smoke_aux)
        else:
            fn = _fn

        if isinstance(solver, optx.OptaxMinimiser):
            context = jax.numpy_dtype_promotion("standard")
        else:
            context = contextlib.nullcontext()
        with context:
            optx_argmin = optx.least_squares(
                fn, solver, init, has_aux=has_aux, args=args, max_steps=10_000, throw=False
            ).value
        out = fn(optx_argmin, args)
        if has_aux:
            residual, _ = out
        else:
            residual = out
        optx_min = jtu.tree_reduce(
            lambda x, y: x + y, jtu.tree_map(lambda x: jnp.sum(x**2), residual)
        )
>       assert tree_allclose(optx_min, minimum, atol=atol, rtol=rtol)
E       assert Array(False, dtype=bool)
E        +  where Array(False, dtype=bool) = tree_allclose(Array(0.12993518, dtype=float64), Array(0., dtype=float64, weak_type=True), atol=0.0001, rtol=0.0001)

tests/test_least_squares.py:53: AssertionError

```

**Comment by johannahaffner** [comment]
So this is a weird one! I took a look, running the tests from current `main` and from `952d0d2`, the version bump commit for the last release, and the last release of `optax`. Equinox is installed from `git`, since this is required for `0.7.1` compatibility. 

With this setup, this test fails on 0.7.1 and passes on 0.7.0. Nothing in our solver logic has changed, nothing in `optax` logic has has changed since this is just installed from PyPI. I replaced our `filter_value_and_grad`, which is used to get the gradient in the `OptaxMinimiser` with the pure JAX equivalent (`jax.value_and_grad`) and get the same test failure, so we must be getting a different value for the gradient of the same function somehow, as of JAX 0.7.1. 

The funny thing is that I can get this test to pass - by accessing the gradient values in `OptaxMinimiser.step` with a `jax.debug.print`. That tells me we're probably looking at some kind of JAX internal bug, perhaps related to memory management (?). I'm going to see if I can get write up a decent JAX-only MWE later tonight :)

**Comment by johannahaffner** [comment]
A plethora of little tweaks to the code restores the expected behaviour, but we did not identify one specific change + have concluded that we're running into some numerical pathology of the Rosenbrock problem here. 

What convinces me of that is that other solvers - with entirely different logic - also start to struggle from nearby starting values, and that adding tiny constants in some places also helps "fix" this. I've opened a PR that slightly modifies the initial value to avoid running into this particular trajectory.
